### PR TITLE
Improve CostAnalyzer

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: src/ir/cost.h
Comment (MaxGraey) [inline src/ir/cost.h:465]: @kripken I think `cmov` and `jmp` should have same costs, so I switch base value to `1` similar to `visitIf`

**Comment by MaxGraey** [comment]
@kripken @tlively could you take a look at this?

**Comment by MaxGraey** [comment]
What do you think about `ignoreCondition` for `CostAnalyzer`? For example when opt pass should decide which is better `If` or `Select` cost of `condition` shouldn't take into account in my opinion

File: src/ir/cost.h
Comment (kripken) [inline src/ir/cost.h:71]: Thanks for noticing these missing things. Good to fix them.

I think we may want to do a larger rewrite here eventually, to automatically visit children, but not before the boilerplate-removing PRs stuff lands.

File: src/ir/cost.h
Comment (kripken) [reply src/ir/cost.h:478]: Sounds reasonable.

File: src/ir/cost.h
Comment (kripken) [inline src/ir/cost.h:487]: Some day we may want to have a "predictability" factor too, as copy (and call) may be much more, we can't tell. But probably not worth it in the near future.

**Review by kripken — APPROVED** [review]
Nice, thanks!

**Comment by kripken** [comment]
> What do you think about add ignoreCondition param for CostAnalyzer? For example when opt pass should decide which is better If or Select cost of condition shouldn't take into account in my opinion

That works out already, doesn't it? The condition will be on both sides, so it cancels out,

```
ifCost <?> selectCost
conditionCost + otherIfCosts <?> conditionCost + otherSelectCosts
otherIfCosts <?> otherSelectCosts
```

**Comment by kripken** [comment]
Oh, do you mean as an optimization for compile time? That might help a small amount - not sure it would be worth the complexity.

**Comment by MaxGraey** [comment]
The main motivation of ignoring condition is normalize this behaviour:
```ts
function test1(x: i32): bool {
  if (x % 2) return true
  else return false
}
```
and
```ts
function test2(x: i32): bool {
  if (x & 1) return true
  else return false
}
```

First one produce:
```wat
(func $test1 (param $0 i32) (result i32)
  local.get $0
  i32.const 2
  i32.rem_s
  if (result i32)
   i32.const 1
  else
   i32.const 0
  end
 )
```
and second:
```wat
(func $test2 (param $0 i32) (result i32)
  local.get $0
  i32.const 1
  i32.and
 )
```
But I expected first one could be:
```wat
(func $test1 (param $0 i32) (result i32)
  i32.const 1
  i32.const 0
  local.get $0
  i32.const 2
  i32.rem_s
  select
 )
```

**Comment by kripken** [comment]
Hmm, that does look like a bug. If it's not fixed by this PR, then it should be optimized in `selectify()` in `RemoveUnusedBrs`. When optimizing for size, it should definitely turn an if like that into a select. `selectify()` is a pretty short function, so should be easy to debug to see what goes wrong.

**Comment by MaxGraey** [comment]
I guess with `rem_s` it just exceeded this threshold:
https://github.com/WebAssembly/binaryen/blob/master/src/passes/RemoveUnusedBrs.cpp#L66

Oh, I see this:
https://github.com/WebAssembly/binaryen/blob/master/src/passes/RemoveUnusedBrs.cpp#L329

Seems it use `br->condition` which reject selectify later

**Comment by kripken** [comment]
I don't think that's the right call. See

https://github.com/WebAssembly/binaryen/blob/f1f2843699f5fd7b87dcefe00ce1eb8f72b37465/src/passes/RemoveUnusedBrs.cpp#L935-L939

which is in `selectify` - that doesn't look at the condition, but at the arms. So something else must be going wrong if it doesn't optimize your case.

(In the call you linked to, the code is correct I think - the condition there is of the `br_if` inside the `if`, so the inner condition may not execute, and must be measured.)

**Comment by MaxGraey** [comment]
Yeah, I tried different approaches and thresholds but result always the same for `-O4`. Test case:
```wat
(func $test (param $x i32) (result i32)
  (if (result i32)
    (i32.rem_s
      (local.get $x)
      (i32.const 4)
    )
    (i32.const 1)
    (i32.const 0)
  )
 )
```

**Comment by MaxGraey** [comment]
So it seems problem only with `-O4` mode

**Comment by kripken** [comment]
Interesting, we should be able to optimize that simple case...

I think I found the problem now, see https://github.com/WebAssembly/binaryen/pull/3310 That fixes the testcase here (kind of embarrassing, as this is pretty old code...)

**Pull request merged**
